### ISSUE TITLE
AsyncCompletion for ManagedFunction

### ISSUE THREAD

**Issue by sagenschneider** [sagenschneider/OfficeFloor#216]
Provide following:

 `AsyncCompletion ManagedFunctionContext.startAsync()`

With `AyncCompletion` having the interface:

  `AsyncCompletion.complete(Throwable)`

This will allow undertaking asynchronous operations within the `ManagedFunction`.  This should tie into the same latch as the `AsyncContext` to time out if takes too long (to progress the functionality).

**Comment by sagenschneider** [comment]
AsyncCompletion interface should allow input Runnable (with potential to throw exception).  This will allow synchronising back to the ThreadState to update the ManagedObject instances (to enable thread safe updates when likely different thread completing the asynchronous operation).

```
public interface AsyncCompletion {
  void complete(Throwable possibleFailure, Runnable threadSafeUpdateManagedObjects);
}
```
This will then allow following interface injected to methods:

```
@AsyncComplete
public interface AnyName {
  void justFinished();
  void failedFinish(Throwable failure);
  void successfulFinish(Runnable updateManagedObjects);
  void completeFinish(Throwable failure, Runnable updateManagedObjects);
}
```
For each method, it will create a new startAsync that needs to be called to finish.  This will allow ClassManagedFunctionSource to create proxies to remove need on OfficeFloor direct implementations.

**Comment by sagenschneider** [comment]
Implemented as AsynchronousFlow.  This is to match naming concepts of Flow.  Best way to think of it: asynchronous flow outside the OfficeFloor managed flows.

**Comment by sagenschneider** [comment]
Resolves #258
